uspace/drv/bus/usb/ohci/hc.c

```diff
@@ -288,38 +284,51 @@ errno_t ohci_hc_status(hcd_t *hcd, uint32_t *status)
  * @param[in] batch Batch representing the transfer.
  * @return Error code.
  */
-errno_t ohci_hc_schedule(hcd_t *hcd, usb_transfer_batch_t *batch)
+errno_t ohci_hc_schedule(usb_transfer_batch_t *batch)
 {
-	assert(hcd);
-	hc_t *instance = hcd_get_driver_data(hcd);
-	assert(instance);
+	assert(batch);
+
+	ohci_bus_t *bus = (ohci_bus_t *) endpoint_get_bus(batch->ep);
+	hc_t *hc = bus->hc;
+	assert(hc);
 
 	/* Check for root hub communication */
-	if (batch->ep->address == ohci_rh_get_address(&instance->rh)) {
-		usb_log_debug("OHCI root hub request.\n");
-		return ohci_rh_schedule(&instance->rh, batch);
+	if (batch->target.address == ohci_rh_get_address(&hc->rh)) {
+		usb_log_debug("OHCI root hub request.");
+		return ohci_rh_schedule(&hc->rh, batch);
 	}
+
+	endpoint_t *ep = batch->ep;
+	ohci_endpoint_t * const ohci_ep = ohci_endpoint_get(ep);
 	ohci_transfer_batch_t *ohci_batch = ohci_transfer_batch_get(batch);
-	if (!ohci_batch)
-		return ENOMEM;
+	int err;
+
+	fibril_mutex_lock(&hc->guard);
+	if ((err = endpoint_activate_locked(ep, batch))) {
+		fibril_mutex_unlock(&hc->guard);
+		return err;
+	}
+
+	if ((err = ohci_transfer_batch_prepare(ohci_batch)))
+		return err;
```

Comment:
Missing mutex unlock?

Reply:
No, it's few lines below. OHCI is designed so wrong, that it needed to lock the endpoint before the transfer is prepared. In short: it has to reuse the last TD of previous transfer as the first one.

But thank's for the comment, because we overcome this and forgot that now it's not needed :)

Reply:
I think this comment was re. the error path on line 313.

Reply:
Oh, of course you're right then. Will be solved by moving out of the critical section.